source/common/router/filter_config_discovery_impl.cc

```diff
@@ -0,0 +1,197 @@
+  for (auto* provider : filter_config_providers_) {
+    provider->onConfigUpdate(factory_callback, version_info);
+  }
+}
+
+void FilterConfigSubscription::onConfigUpdate(
+    const std::vector<Envoy::Config::DecodedResourceRef>& added_resources,
+    const Protobuf::RepeatedPtrField<std::string>& removed_resources, const std::string&) {
+  if (!removed_resources.empty()) {
+    ENVOY_LOG(error,
+              "Server sent a delta FilterConfigDS update attempting to remove a resource (name: "
+              "{}). Ignoring.",
+              removed_resources[0]);
+  }
+  if (!added_resources.empty()) {
+    onConfigUpdate(added_resources, added_resources[0].get().version());
+  }
+}
+
+void FilterConfigSubscription::onConfigUpdateFailed(Envoy::Config::ConfigUpdateFailureReason reason,
+                                                    const EnvoyException*) {
+  ENVOY_LOG(debug, "Updating filter config {} failed due to {}", filter_config_name_, reason);
+  // Make sure to make progress in case the control plane is temporarily failing.
+  local_init_target_.ready();
+}
+
+FilterConfigSubscription::~FilterConfigSubscription() {
+  // If we get destroyed during initialization, make sure we signal that we "initialized".
+  local_init_target_.ready();
+  // Remove the subscription from the provider manager.
+  filter_config_provider_manager_.subscriptions_.erase(subscription_id_);
+}
+
+std::shared_ptr<FilterConfigSubscription> FilterConfigProviderManagerImpl::getSubscription(
+    const envoy::config::core::v3::ConfigSource& config_source, const std::string& name,
+    Server::Configuration::FactoryContext& factory_context, const std::string& stat_prefix) {
+  // FilterConfigSubscriptions are unique based on their config source and filter config name
+  // combination.
+  const std::string subscription_id = absl::StrCat(MessageUtil::hash(config_source), ".", name);
+  auto it = subscriptions_.find(subscription_id);
+  if (it == subscriptions_.end()) {
+    auto subscription = std::make_shared<FilterConfigSubscription>(
+        config_source, name, factory_context, stat_prefix, *this, subscription_id);
+    subscriptions_.insert({subscription_id, std::weak_ptr<FilterConfigSubscription>(subscription)});
+    return subscription;
+  } else {
+    auto existing = it->second.lock();
+    RELEASE_ASSERT(existing != nullptr,
+                   absl::StrCat("Cannot find subscribed filter config resource ", name));
+    return existing;
+  }
+}
+
+FilterConfigProviderPtr FilterConfigProviderManagerImpl::createDynamicFilterConfigProvider(
+    const envoy::config::core::v3::ConfigSource& config_source,
+    const std::string& filter_config_name, bool require_terminal,
+    Server::Configuration::FactoryContext& factory_context, const std::string& stat_prefix,
+    bool apply_without_warming) {
+  auto subscription =
+      getSubscription(config_source, filter_config_name, factory_context, stat_prefix);
+  if (!apply_without_warming) {
```

Comment:
Can you add some comments on when we hit this case? I.e. this is because we're applying some default filter config. Although, if this is default config, why can't this be treated like a static config provider?

Reply:
Done. Default config means default for version 0, so it can be still changed with xDS.